modules/support/src/dex.rs

```diff
@@ -72,24 +72,150 @@ pub trait DEXManager<AccountId, CurrencyId, Balance> {
 	) -> sp_std::result::Result<(Balance, Balance), DispatchError>;
 }
 
-pub trait StableAssetDEX<AccountId, Balance, CurrencyId> {
-	fn get_best_price_pool(
+pub trait Swap<AccountId, Balance, CurrencyId> {
+	fn get_swap_amount(
 		supply_currency_id: CurrencyId,
 		target_currency_id: CurrencyId,
 		limit: SwapLimit<Balance>,
-	) -> Option<(StableAssetPoolId, PoolTokenIndex, PoolTokenIndex)>;
+	) -> Option<(Balance, Balance)>;
 
 	fn swap(
 		who: &AccountId,
-		pool_id: StableAssetPoolId,
-		supply_asset_index: PoolTokenIndex,
-		target_asset_index: PoolTokenIndex,
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
 		limit: SwapLimit<Balance>,
 	) -> sp_std::result::Result<(Balance, Balance), DispatchError>;
 }
 
+pub struct SpecificJointsSwap<Dex, Joints>(sp_std::marker::PhantomData<(Dex, Joints)>);
+
+impl<AccountId, Balance, CurrencyId, Dex, Joints> Swap<AccountId, Balance, CurrencyId>
+	for SpecificJointsSwap<Dex, Joints>
+where
+	Dex: DEXManager<AccountId, Balance, CurrencyId>,
+	Joints: Get<Vec<Vec<CurrencyId>>>,
+	Balance: Clone,
+{
+	fn get_swap_amount(
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
+		limit: SwapLimit<Balance>,
+	) -> Option<(Balance, Balance)> {
+		<Dex as DEXManager<AccountId, Balance, CurrencyId>>::get_best_price_swap_path(
+			supply_currency_id,
+			target_currency_id,
+			limit,
+			Joints::get(),
+		)
+		.map(|(_, supply_amount, target_amount)| (supply_amount, target_amount))
+	}
+
+	fn swap(
+		who: &AccountId,
+		supply_currency_id: CurrencyId,
+		target_currency_id: CurrencyId,
+		limit: SwapLimit<Balance>,
+	) -> sp_std::result::Result<(Balance, Balance), DispatchError> {
+		let path = <Dex as DEXManager<AccountId, Balance, CurrencyId>>::get_best_price_swap_path(
+			supply_currency_id,
+			target_currency_id,
+			limit.clone(),
+			Joints::get(),
+		)
+		.ok_or(DispatchError::Other("Cannot swap"))?
```

Comment:
better to avoid `DispatchError::Other`

Could do something like this 
https://github.com/AcalaNetwork/Acala/blob/master/primitives/src/bonding/error.rs
https://github.com/AcalaNetwork/Acala/blob/master/modules/earning/src/lib.rs#L265